File: python_email_service/debug_database.py
#!/usr/bin/env python3
"""
Debug Database - Check what's in the users table
"""

import sys
import os
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

import psycopg2
from psycopg2.extras import RealDictCursor

# Database connection
DATABASE_URL = "postgresql://neondb_owner:npg_PZae5A9gHpbY@ep-fragrant-violet-ad4xrlr3-pooler.c-2.us-east-1.aws.neon.tech/neondb?sslmode=require"

def debug_database():
    """Debug what's in the database"""
    print("🔍 DATABASE DEBUG")
    print("=" * 50)
    
    try:
        conn = psycopg2.connect(DATABASE_URL)
        cursor = conn.cursor(cursor_factory=RealDictCursor)
        
        # Check if users table exists
        print("\n1. Checking if users table exists...")
        cursor.execute("""
            SELECT EXISTS (
                SELECT FROM information_schema.tables 
                WHERE table_name = 'users'
            )
        """)
        users_table_exists = cursor.fetchone()['exists']
        print(f"   Users table exists: {users_table_exists}")
        
        if not users_table_exists:
            print("❌ Users table doesn't exist!")
            return
        
        # Check total users
        print("\n2. Checking total users...")
        cursor.execute("SELECT COUNT(*) as total FROM users")
        total_users = cursor.fetchone()['total']
        print(f"   Total users: {total_users}")
        
        # Check users with emails
        print("\n3. Checking users with emails...")
        cursor.execute("""
            SELECT COUNT(*) as total 
            FROM users 
            WHERE email IS NOT NULL AND email != ''
        """)
        users_with_emails = cursor.fetchone()['total']
        print(f"   Users with emails: {users_with_emails}")
        
        # Show all users
        print("\n4. All users in database:")
        cursor.execute("""
            SELECT user_id, name, email, calorie_goal, protein_goal, created_at
            FROM users 
            ORDER BY user_id
        """)
        users = cursor.fetchall()
        
        if users:
            for user in users:
                email_status = "✅" if user['email'] else "❌"
                print(f"   {email_status} ID: {user['user_id']}, Name: {user['name']}, Email: {user['email']}")
                print(f"      Goals: {user['calorie_goal']}cal, {user['protein_goal']}g protein")
        else:
            print("   No users found!")
        
        # Check email_preferences table
        print("\n5. Checking email_preferences table...")
        cursor.execute("""
            SELECT EXISTS (
                SELECT FROM information_schema.tables 
                WHERE table_name = 'email_preferences'
            )
        """)
        email_prefs_exists = cursor.fetchone()['exists']
        print(f"   Email preferences table exists: {email_prefs_exists}")
        
        if email_prefs_exists:
            cursor.execute("SELECT COUNT(*) as total FROM email_preferences")
            total_prefs = cursor.fetchone()['total']
            print(f"   Total email preferences: {total_prefs}")
        
        # Test the exact query from email service
        print("\n6. Testing email service query...")
        if email_prefs_exists:
            cursor.execute("""
                SELECT 
                    u.user_id,
                    u.name,
                    u.email,
                    u.calorie_goal,
                    u.protein_goal,
                    u.created_at,
                    COALESCE(ep.daily_reminders_enabled, true) as daily_reminders_enabled,
                    COALESCE(ep.achievement_notifications_enabled, true) as achievement_notifications_enabled,
                    ep.reminder_time
                FROM users u
                LEFT JOIN email_preferences ep ON u.user_id = ep.user_id
                WHERE u.email IS NOT NULL 
                AND u.email != ''
                AND u.email != 'user@example.com'
                AND COALESCE(ep.daily_reminders_enabled, true) = true
                ORDER BY u.created_at DESC
            """)
        else:
            cursor.execute("""
                SELECT 
                    u.user_id,
                    u.name,
                    u.email,
                    u.calorie_goal,
                    u.protein_goal,
                    u.created_at,
                    true as daily_reminders_enabled,
                    true as achievement_notifications_enabled,
                    '09:00:00' as reminder_time
                FROM users u
                WHERE u.email IS NOT NULL 
                AND u.email != ''
                AND u.email != 'user@example.com'
                ORDER BY u.created_at DESC
            """)
        
        email_users = cursor.fetchall()
        print(f"   Users found by email service query: {len(email_users)}")
        
        for user in email_users:
            print(f"   📧 {user['name']} ({user['email']}) - {user['calorie_goal']}cal, {user['protein_goal']}g")
        
        cursor.close()
        conn.close()
        
        print(f"\n📊 SUMMARY:")
        print(f"   Total users: {total_users}")
        print(f"   Users with emails: {users_with_emails}")
        print(f"   Users for email service: {len(email_users)}")
        
        if len(email_users) == 0:
            print("\n❌ PROBLEM: No users found for email service!")
            print("   Possible issues:")
            print("   1. No users have email addresses")
            print("   2. All emails are empty or 'user@example.com'")
            print("   3. Email preferences are disabled")
        else:
            print(f"\n✅ SUCCESS: Found {len(email_users)} users for email service!")
        
    except Exception as e:
        print(f"❌ Database error: {e}")

if __name__ == "__main__":
    debug_database()

File: debug-database.js
#!/usr/bin/env node
/**
 * Debug Database Direct Query
 * Check what's actually in the database
 */

import { neon } from "@neondatabase/serverless"

const DATABASE_URL = "postgresql://neondb_owner:npg_PZae5A9gHpbY@ep-fragrant-violet-ad4xrlr3-pooler.c-2.us-east-1.aws.neon.tech/neondb?sslmode=require"
const sql = neon(DATABASE_URL)

async function debugDatabase() {
    console.log('🔍 DEBUGGING DATABASE DIRECTLY');
    console.log('=' .repeat(60));

    try {
        // Check users
        console.log('\n👥 USERS:');
        const users = await sql`SELECT user_id, name, calorie_goal, protein_goal FROM users ORDER BY user_id`
        users.forEach(user => {
            console.log(`   User ${user.user_id}: ${user.name} (${user.calorie_goal} cal, ${user.protein_goal}g protein)`);
        });

        // Check meals for today
        const today = new Date().toISOString().split('T')[0];
        console.log(`\n🍽️ MEALS FOR TODAY (${today}):`);
        const meals = await sql`
            SELECT user_id, meal_text, calories, protein, meal_date 
            FROM meals 
            WHERE meal_date = ${today}
            ORDER BY user_id, created_at
        `
        meals.forEach(meal => {
            console.log(`   User ${meal.user_id}: ${meal.meal_text} (${meal.calories} cal, ${meal.protein}g protein)`);
        });

        // Check user achievements
        console.log('\n🏆 USER ACHIEVEMENTS:');
        const userAchievements = await sql`
            SELECT 
                ua.user_id,
                a.name,
                ua.current_progress,
                ua.is_earned,
                ua.updated_at
            FROM user_achievements ua
            JOIN achievements a ON ua.achievement_id = a.achievement_id
            WHERE ua.user_id IN (6, 8) AND a.name LIKE '%Daily%'
            ORDER BY ua.user_id, a.name
        `
        userAchievements.forEach(ua => {
            console.log(`   User ${ua.user_id}: ${ua.name} = ${ua.current_progress} (${ua.is_earned ? 'EARNED' : 'not earned'}) - Updated: ${ua.updated_at}`);
        });

        // Check achievements table
        console.log('\n📋 ACHIEVEMENTS DEFINITIONS:');
        const achievements = await sql`
            SELECT achievement_id, name, achievement_type, target_value, target_unit
            FROM achievements 
            WHERE name LIKE '%Daily%'
            ORDER BY name
        `
        achievements.forEach(a => {
            console.log(`   ${a.achievement_id}: ${a.name} (${a.achievement_type}, target: ${a.target_value} ${a.target_unit})`);
        });

        // Test the API query directly
        console.log('\n🔧 TESTING API QUERY FOR USER 8:');
        const apiResult = await sql`
            SELECT 
                a.achievement_id,
                a.name,
                a.achievement_type,
                a.target_value,
                a.target_unit,
                COALESCE(ua.is_earned, false) as is_earned,
                COALESCE(ua.current_progress, 0) as current_progress,
                ua.earned_at,
                CASE 
                  WHEN ua.is_earned THEN 100
                  WHEN a.achievement_type = 'daily' THEN LEAST(100, COALESCE(ua.current_progress, 0) * 100)
                  WHEN a.target_value > 0 THEN LEAST(100, (COALESCE(ua.current_progress, 0) / a.target_value) * 100)
                  ELSE 0
                END as progress_percentage
            FROM achievements a
            LEFT JOIN user_achievements ua ON a.achievement_id = ua.achievement_id AND ua.user_id = 8
            WHERE a.is_active = true AND a.name LIKE '%Daily%'
            ORDER BY a.name
        `
        
        apiResult.forEach(result => {
            console.log(`   ${result.name}:`);
            console.log(`      Current Progress: ${result.current_progress}`);
            console.log(`      Progress Percentage: ${result.progress_percentage}%`);
            console.log(`      Achievement Type: ${result.achievement_type}`);
            console.log(`      Target Value: ${result.target_value}`);
        });

    } catch (error) {
        console.error('❌ Database error:', error);
    }
}

debugDatabase().catch(console.error);
